### PR TITLE
CP-17795: Implement gfs2 Lun resize

### PR DESCRIPTION
**Pull request by geosharath**
Signed-off-by: Sharath Babu <Sharath.Babu@citrix.com>

### CODE REVIEW

File: volume/org.xen.xapi.storage.gfs2/sr.py [hunk 1]
```diff
@@ -513,7 +552,56 @@ def detach(self, dbg, sr):
...
+                        call(dbg, ["gfs2_grow", mountpoint_root + "dev/" + vg_name +"/gfs2"])
+
+                        #inform other node about LUN resize
+                        inventory = xcp.environ.readInventory()
+                        session = XenAPI.xapi_local()
+                        session.xenapi.login_with_password("root", "")
+                        this_host = session.xenapi.host.get_by_uuid(
+                            inventory.get("INSTALLATION_UUID"))
+
+                        for host in session.xenapi.host.get_all():
```
Comment (stefanopanella) [inline volume/org.xen.xapi.storage.gfs2/sr.py]: I think the refresh should go just after the lvresize, before gfsgrow

Comment (geosharath) [inline volume/org.xen.xapi.storage.gfs2/sr.py]: yes, Ill make that change.

File: volume/org.xen.xapi.storage.gfs2/sr.py [hunk 2]
```diff
@@ -26,6 +26,45 @@
 # For a block device /a/b/c, we will mount it at <mountpoint_root>/a/b/c
 mountpoint_root = "/var/run/sr-mount/"
 DLM_REFDIR = "/var/run/sr-ref"
+VG_FREE_SPACE_THRESHOLD = 0
+
+def exec_cmd(cmd):
```
Comment (stefanopanella) [inline volume/org.xen.xapi.storage.gfs2/sr.py]: Why do you need to define a new method to call an executable?
Could you not just use call() as you are using later for lvresize ?

File: volume/org.xen.xapi.storage.gfs2/sr.py [hunk 3]
```diff
@@ -513,7 +552,56 @@ def detach(self, dbg, sr):
...
+                log.debug("Free space (%s) detected in VG, expanding gfs2 LV." %str(stats['freespace']))
+                opq = urlparse.urlparse(sr).path
+                try:
+                    with Lock(opq, 'gl', gfs2.Callbacks()):
+                        # extend lv
+                        call(dbg, ["lvextend", "-l+100%FREE", lv_name, "--config", "global{metadata_read_only=0}"])
+                        # grow gfs2
+                        call(dbg, ["gfs2_grow", mountpoint_root + "dev/" + vg_name +"/gfs2"])
+
+                        #inform other node about LUN resize
```
Comment (stefanopanella) [inline volume/org.xen.xapi.storage.gfs2/sr.py]: I think the "nform other node about LUN resize" should go just after the lvresize, before gfsgrow

File: volume/org.xen.xapi.storage.gfs2/sr.py (hunk 2 above)
Comment (geosharath) [inline volume/org.xen.xapi.storage.gfs2/sr.py]: wasn't sure if the custom call will return the output, will reuse it if returns the same.

File: volume/org.xen.xapi.storage.gfs2/sr.py [hunk 4]
```diff
@@ -513,7 +552,56 @@ def detach(self, dbg, sr):
...
+            # incase if LUN is resized pv size will get updated
+            call(dbg, ["/usr/sbin/pvresize" , pv_name, "--config", "global{metadata_read_only=0}"])
+
+            # if pv was expanded, this will reflect as freespace
+            # in the associated volume group, only then we need to expand gfs2 lv
+
+            stats = vg_stats(dbg,vg_name)
+            if stats['freespace'] > VG_FREE_SPACE_THRESHOLD:
+                log.debug("Free space (%s) detected in VG, expanding gfs2 LV." %str(stats['freespace']))
+                opq = urlparse.urlparse(sr).path
```
Comment (stefanopanella) [inline volume/org.xen.xapi.storage.gfs2/sr.py]: Not sure if opq and Lock would work here

Comment (geosharath) [inline volume/org.xen.xapi.storage.gfs2/sr.py]: yes, I forgot to add the lock class to the file while creating the PR. will try util.lock_file as you suggested

**Comment by stefanopanella** [comment]
Thanks Sharat, it looks good now, please integrate this in trunk

**Comment by geosharath** [comment]
merged to master

**Pull request closed without merging**